### PR TITLE
Fix #12860, 2bada59: Timetable does not show vehicles as early

### LINKED ISSUE #12860: [Bug]: Timetable status doesn't say if a vehicle is early
### Version of OpenTTD

Nightly (eeed824)

### Expected result

When a vehicle arrives early, its status should say "N days early".

### Actual result

The vehicle says "running on time".

### Steps to reproduce

1. Create a vehicle traveling between two stations and autofill its timetable.
2. Change one running time to a larger number, say from 9 days to 30 days.
3. When it arrives at a station, it will wait until its departure time, but the status will say "running on time" and not "N days early".

### PR DESCRIPTION
**Pull request by ivpravdin**
## Motivation / Problem

Fix #12860. Timetable does not show vehicles as early, only as on time. If the vehicle's travel time is greater than the actual time it takes to reach the destination, the vehicle will arrive early but the timetable will display it as "running on time". 

## Description

This bug was fixed by changing VehicleIsAboveLatenessThreshold() to only accept how many ticks the vehicle is late, not the entire vehicle instance. This allows to pass an absolute value of number of ticks and therefore determine whether the vehicle is on time or late/early.

## Limitations

Limitation (unlikely): if in the future vehicle's late status will have to be determined by anything else other than number of ticks, new parameters will have to be added to VehicleIsAboveLatenessThreshold().

### CODE REVIEW

File: src/timetable_gui.cpp [hunk 1]
```diff
@@ -87,19 +87,19 @@ static inline TimerGameTick::Ticks TicksPerTimetableUnit()
...
  * @param round_to_day When using ticks, if we should round up to the nearest day.
  * @return True if the vehicle is later than the threshold.
  */
-bool VehicleIsAboveLatenessThreshold(const Vehicle *v, bool round_to_day)
+bool VehicleIsAboveLatenessThreshold(TimerGameTick::Ticks ticks, bool round_to_day)
 {
 	switch (_settings_client.gui.timetable_mode) {
 		case TimetableMode::Days:
-			return v->lateness_counter > Ticks::DAY_TICKS;
+			return ticks > Ticks::DAY_TICKS;
```
Comment (2TallTyler) [inline src/timetable_gui.cpp:98]: Thanks for the contribution!

Why change what is passed to the function, if all that's needed is to take the absolute value of the lateness counter?

If I'm not mistaken, this fix could be confined to `VehicleIsAboveLatenessThreshold()` by simply changing each switch case as below.

```suggestion
			return abs(v->lateness_counter) > Ticks::DAY_TICKS;
```

File: src/timetable_gui.cpp [hunk 2]
```diff
@@ -87,19 +87,19 @@ static inline TimerGameTick::Ticks TicksPerTimetableUnit()
 
 /**
  * Determine if a vehicle should be shown as late, depending on the timetable display setting.
- * @param v The vehicle in question.
+ * @param ticks How many ticks the vehicle is late.
  * @param round_to_day When using ticks, if we should round up to the nearest day.
  * @return True if the vehicle is later than the threshold.
```
Comment (2TallTyler) [inline src/timetable_gui.cpp]:
```suggestion
 * @return True if the vehicle is outside the "on time" threshold, either early or late.
```

File: src/timetable_gui.cpp [hunk 3]
```diff
@@ -87,19 +87,19 @@ static inline TimerGameTick::Ticks TicksPerTimetableUnit()
 
 /**
  * Determine if a vehicle should be shown as late, depending on the timetable display setting.
```
Comment (2TallTyler) [inline src/timetable_gui.cpp]:
```suggestion
 * Determine if a vehicle should be shown as late or early, using a threshold depending on the timetable display setting.
```

File: src/timetable_gui.cpp (hunk 1 above)
Comment (ivpravdin) [reply src/timetable_gui.cpp:98]: Hi, thank you for your suggestion. The problem here is that is that in this line:
`504: bool show_late = this->show_expected && VehicleIsAboveLatenessThreshold(v->lateness_counter, true);` it is necessary to actually determine whether the vehicle is late (and not early). With you suggestion `show_late` would be true for early vehicles as well. I've tried it the way you suggested and that's the issue I ran into. I think it makes sense to make it possible to determine if vehicle is both later/early or just late.

File: src/timetable_gui.cpp [hunk 4]
```diff
@@ -87,19 +87,19 @@ static inline TimerGameTick::Ticks TicksPerTimetableUnit()
...
- * @param v The vehicle in question.
+ * @param ticks How many ticks the vehicle is late.
```
Comment (2TallTyler) [inline src/timetable_gui.cpp]:
```suggestion
 * @param ticks The number of ticks that the vehicle is late or early.
```

File: src/timetable_gui.cpp (hunk 1 above)
Comment (2TallTyler) [reply src/timetable_gui.cpp:98]: Ah, thanks for the explanation. You are correct. 🙂 

Do you mind applying my comment suggestions? Then this is ready for an approval. 😃

**Review by 2TallTyler — APPROVED** [review]

**Comment by 2TallTyler** [comment]
Checks are failing due to #12881, nothing you can do until that is fixed. 🙂

**Pull request merged**